engine/storage/snapshot/src/org/apache/cloudstack/storage/snapshot/SnapshotServiceImpl.java

```diff
@@ -300,17 +300,24 @@ protected Void copySnapshotAsyncCallback(AsyncCallbackDispatcher<SnapshotService
         CopyCommandResult result = callback.getResult();
         SnapshotInfo destSnapshot = context.destSnapshot;
         SnapshotObject srcSnapshot = (SnapshotObject)context.srcSnapshot;
+        Object payload = srcSnapshot.getPayload();
+        CreateSnapshotPayload createSnapshotPayload = (CreateSnapshotPayload)payload;
         AsyncCallFuture<SnapshotResult> future = context.future;
         SnapshotResult snapResult = new SnapshotResult(destSnapshot, result.getAnswer());
         if (result.isFailed()) {
             try {
-                destSnapshot.processEvent(Event.OperationFailed);
-                //if backup snapshot failed, mark srcSnapshot in snapshot_store_ref as failed also
-                srcSnapshot.processEvent(Event.DestroyRequested);
-                srcSnapshot.processEvent(Event.OperationSuccessed);
-
-                srcSnapshot.processEvent(Snapshot.Event.OperationFailed);
-                _snapshotDao.remove(srcSnapshot.getId());
+                if (createSnapshotPayload.getSeparateBackup()) {
+                    destSnapshot.processEvent(Event.OperationFailed);
+                    throw new NoTransitionException("Failed in creating backup retrying");
```

Comment:
The purpose in throwing the exception explicitly is to handle the multiple retries in creating backup, in case of failure. And the srcSnapshot is handled properly, its even destroyed after the max retry attempts.